### PR TITLE
Feature/web server

### PR DESCRIPTION
**Pull request by thiagonache**

### CODE REVIEW

File: cmd/linksws/main.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package main
+
+import (
+	"links"
+	"log"
+)
+
+func main() {
+	log.Fatal(links.ServeHTTP())
```
Comment (bitfield) [inline cmd/linksws/main.go]: Maybe `ListenAndServe`, for symmetry with `http`?

File: webserver.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	"net/http"
+	"os"
+	"strconv"
+)
+
+func webServerCheck(w http.ResponseWriter, r *http.Request) {
+	queryString := r.URL.Query()
+	qsSite := queryString["site"]
+	if len(qsSite) == 0 {
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (bitfield) [inline webserver.go]: Perhaps `http.Error` for short?

File: webserver.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			fmt.Fprintln(w, `{"err":"cannot convert no-recursion to boolean"`)
+			return
+		}
+	}
+	output := io.Discard
+	qsOutput := queryString["output"]
+	if len(qsOutput) > 0 {
+		output = os.Stdout
```
Comment (bitfield) [inline webserver.go:46]: Then what's `qsOutput` for?

File: webserver.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	results, err := Check(qsSite, WithStdout(output), WithDebug(debug), WithNoRecursion(noRecursion))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(w, err.Error())
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(results)
```
Comment (bitfield) [inline webserver.go]: `Encode` returns error.

File: webserver.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+func WebServerHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
```
Comment (bitfield) [inline webserver.go]: We don't need a switch statement if there's just one non-default case!

File: webserver.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	default:
+		w.WriteHeader(http.StatusMethodNotAllowed)
+	}
+}
+
+func ServeHTTP() error {
+	router := http.NewServeMux()
+	handlerCheck := http.HandlerFunc(WebServerHandler)
+	router.Handle("/check", handlerCheck)
+	return http.ListenAndServe(":8080", router)
```
Comment (bitfield) [inline webserver.go]: This uses the default HTTP server, which has no timeouts. Let's create a custom server with the standard CloudFlare config: https://blog.cloudflare.com/exposing-go-on-the-internet/

File: webserver_test.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"links"
+)
+
+func TestCheckWebServer(t *testing.T) {
+	t.Parallel()
+	request, _ := http.NewRequest(http.MethodGet, "/check?site=https://bitfieldconsulting.com&no-recursion=true", nil)
```
Comment (bitfield) [inline webserver_test.go]: We don't ignore errors, even in tests (especially not in tests). And just `http.Get` would work here, wouldn't it?

**Review by bitfield — APPROVED** [review]
Great stuff!

File: cmd/linksws/main.go (hunk 1 above)
Comment (thiagonache) [reply cmd/linksws/main.go]: yep

File: webserver.go (hunk 2 above)
Comment (thiagonache) [reply webserver.go]: not sure what you mean here.

File: webserver.go (hunk 3 above)
Comment (thiagonache) [reply webserver.go:46]: yeah... it's odd... but I need to be able to set output to stdout for troubleshooting.

File: webserver.go (hunk 4 above)
Comment (thiagonache) [reply webserver.go]: fixed

File: webserver.go (hunk 5 above)
Comment (thiagonache) [reply webserver.go]: right! thx

File: webserver.go (hunk 2 above)
Comment (thiagonache) [reply webserver.go]: gotcha!

**Pull request merged**